README.md
# grunt-flexcombo

![](http://gtms01.alicdn.com/tps/i1/T1TgNqFB0bXXbX25fJ-346-77.png)

Grunt-flexcombo 是一款基于NodeJS的轻服务，便携且易于配置。用于淘系环境中的Demo服务和自定义虚机等场景。

author: @拔赤、@陶清

> Grunt-flexcombo 是 [flex-combo](https://npmjs.org/package/flex-combo)的grunt插件版本，[Grunt-Flexcombo 原理介绍](https://speakerdeck.com/lijing00333/grunt-flexcombo)。

## 环境依赖和包的安装

依赖 Grunt 版本`~0.4.1`

安装grunt-flexcombo

	npm install grunt-flexcombo --save-dev

安装后，在Gruntfile.js中载入任务

	grunt.loadNpmTasks('grunt-flexcombo');

## "flexcombo" 任务配置

在`grunt.initConfig()`中添加flexcombo的配置：

	grunt.initConfig({
		flexcombo:{
            server:{
                options:{
                    target:'src/',
                    urls:'/<%= pkg.group %>/<%= pkg.name %>',
                    port:'<%= pkg.port %>',
                    mockPath: '<%= pkg.mockPath %>',//指定本地mock的目录（路径）名，默认为"mock"
                    servlet:'?',
                    separator:',',
                    charset:'utf8',
                    livereload: true		// 指定是否开启 livereload，默认为 false，也可配置为 livereload 的端口号，与 `watch` 任务的配置保持一致
                }
            },
			debug:{
				options: {
					target:'src',
					urls:'/group/project',// 匹配[ag].tbcdn.cn时的访问路径
					port:'80',// flexcombo服务端口
					proxyport:'8080',// 反向代理服务端口
					servlet:'?',//开启combo的字符串
					separator:',',//多文件分隔符
					charset:'utf8',//默认输出文件的编码
					proxyHosts:['a.com','b.com','demo.com'],//本地服务的域名
					startWeinre: false,		// 是否自动调用 weinre, H5 项目可选
					livereload: true,		// 指定是否开启 livereload，默认为 false，也可配置为 livereload 的端口号，与 `watch` 任务的配置保持一致
					weinrePort: 8091,		// 默认的 weinre 运行端口
					// 代理配置
					proxy: {
						interface: {		// 接口 mock 配置
							hosts: ['api.m.taobao.com', 'api.waptest.taobao.com', 'api.test.taobao.com'],// 要代理的接口域名
							script: 'proxy/interface.js'// 代理脚本路径
						},
						webpage: {			// 页面 HTML 区块代理配置
							urls: [/taobao\.com/],//[/www\.baidu\.com/, /taobao\.com/],		// 要代理的页面路径匹配规则（字符串或正则表达式）
							script: 'proxy/webpage.js'	// 代理脚本路径
						}
					},
					webConfigPort: 8082,	// anyproxy web config 页面端口
					webPort: 8002,			// anyproxy 监控页面端口
					socketPort: 8003		// anyproxy 控制台与监控页面通信端口
					
					// 过滤参数
					filter:{ 
						'-min\\.js':'.js',
						/*
						 * 当访问a.com/go/act/demo.php时，实际请求
						 * a.com/demo.html
						 */
						'/go/act/\(.+\\.\)php':'/$1html'
					}
				}
			}
		},
		// flexcombo 需要配合 watch 模块一起使用
		watch: {
			'all': {
				files: ['src/**/*'],
				tasks: [ 'your-build-target' ]
			}
		}
	});

	// 注册 debug 任务
	grunt.registerTask('debug', '开启debug模式', function() {
		task.run(['flexcombo:debug','watch:all']);
	});

<strong>启动服务：</strong>

	sudo grunt debug

<strong>任务说明：</strong>

`flexcombo`任务无法单独运行，需要配合`watch`启动。一份完整的[Gruntfile.js](https://github.com/jayli/slide/blob/master/Gruntfile.js)样例。

## 原理说明

### 1. 服务启动

上例中`sudo grunt debug`将会调用flexcombo服务，会在本地启动两个服务，两个服务分属两个端口`proxyport`（反向代理服务）和`port`（CDN 环境服务）

![](http://gtms01.alicdn.com/tps/i1/T1xHe5FABbXXbRj_sG-542-385.png)

启动服务后，可以直接通过`localhost`或者IP来预览本地文件

	http://localhost/group/project

也可以通过`proxyHosts`中的域名配置来访问：

	http://demo.com/

若要模拟真实环境调试，两种方法（二者取其一），推荐第二种：

1. <del>将cdn配向本机`127.0.0.1 g.tbcdn.cn a.tbcdn.cn`</del>
1. 将浏览器或者设备HTTP代理配置到本机的反向代理服务的端口

在Firefox中设置代理：

![](http://gtms01.alicdn.com/tps/i1/T1KzNFFrxbXXcrSyrN-372-175.png)

在手机终端设置代理方法：

![](http://gtms01.alicdn.com/tps/i1/T1bePRFlVXXXXhb4nD-502-341.png)

然后访问真实页面，配合watch和你Gruntfile.js中的构建命令，可以开始调试，比如在[Gallery中的代码调试](http://gallery.kissyui.com/quickstart)：

![](http://gtms01.alicdn.com/tps/i1/T1.ZhxFvXbXXb0CP6.-596-364.png)

### 2. SSI

该服务支持标准格式的 [SSI include](http://man.chinaunix.net/newsoft/ApacheManual/howto/ssi.html)

	<!--#include path="../src/file-path.html" -->

### 3. AWP 中的标签替换

目前只支持`<!--HTTP:url,utf8:HTTP-->`的解析，其他类型暂时不支持，这个规则不会被[combohtml](https://www.npmjs.org/package/grunt-combohtml)解析

### 4.Mork 数据模拟

服务支持[juicer模板](http://juicer.name)渲染输出，因此在源html中可以直接用juicer语法来写页面，比如源文件:

	<html>
	<body>
		<!-- 定义Mock数据 -->
		<!--#def
			{"list": [
					{"name":" guokai", "show": true},
					{"name":" benben", "show": false},
					{"name":" dierbaby", "show": true}
				]}
		-->
		<ul>
			{@each list as it,index}
				<li>${it.name} (index: ${index})</li>
			{@/each}
		</ul>
	</body>
	</html>

开启服务后浏览器中访问它，将输出：

	<html>
		<body>
			<!-- 定义Mock数据 -->
			<!--#def {"list": [ {"name":" guokai", "show": true}, 
			{"name":" benben","show": false}, 
			{"name":" dierbaby", "show": true} ]} -->
			<ul>
				<li>guokai (index: 0)</li>
				<li>benben (index: 1)</li>
				<li>dierbaby (index: 2)</li>
			</ul>
		</body>
	</html>

即，数据和juicer模板混合输出了正确的结果。

如果源文件中存在Mock数据字段`<!--#def ... -->`，则服务将会解析文件中的juicer模板

> 这个功能主要提供了调试的渠道，写demo时就直接生成了juicer模板，通过[grunt-combohtml](https://github.com/jayli/grunt-combohtml)（另行配置）可以将juicer语法的源文件编译为velocity语法和php语法。

## 配置参数说明

#### options.longPolling

- 类型：Boolean
- 默认值：false

是否采用长连接模式

#### options.target

- 类型：String
- 默认值：''

基于此目录启用本地服务，类似Apache中的`DocumentRoot`，需写相对目录，相对于`Gruntfile.js`所在的目录。

#### options.urls

- 类型：String
- 默认值：''

启用本地服务后，访问到`options.target`目录对应的url里的路径，比如`options`的配置：

	options:{
		target:'src'
		urls:'/a/b/c'
	}

访问URL`http://localhost/a/b/c/demo.html`，将返回`./src/demo.html`。此配置主要用作模拟cdn的真实路径，比如直接访问`http://g.tbcdn.cn/a/b/c/index.js`，将访问`./src/index.js`。

#### options.port

- 类型：String
- 默认值：'80'

flexcombo 服务的端口，默认启动在80端口，启动在80端口可以让你直接绑定Host到本机，若启动在其他端口，则只能用代理工具来辅助访问。

#### options.proxyport

- 类型：String
- 默认值：'8080'

反向代理服务器的端口。反向代理服务器运行有过滤器，如果反向代理匹配`proxyHosts`和`[ag].tbcdn.cn`，则会将请求转发至flexcombo服务，否则去公网取文件。最后一并返回给客户端。

#### options.servlet

- 类型：String
- 默认：'?'

开启combo的前缀。不同的开发环境有不同的combo需求。通过`servlet`,`seperator`两个参数决定。

#### options.separator

- 类型：String
- 默认值：','

Combo的url中多个文件的分隔符。

#### options.charset

- 类型：String
- 默认：'utf8'

默认输出文件的编码，主要用于约束输出assets文件的编码。

#### options.proxyHosts

- 类型：String 或者 Array
- 默认：'' 或者 []

本地服务自定义虚机的域名。除了`a.tbcdn.cn`和`g.tbcdn.cn`这两个默认域名，可以新增自定义域名配置，多个域名用数组表示。

> 自定义域名无须配Host访问，只需配端口为`proxyport`的代理服务即可。

注意：来自`proxyHosts`中配置的域名的url请求路径不会匹配`options.urls`，比如`proxyHosts:'a.com'`，那么访问

	http://a.com/demo.html

将访问本地文件`./src/demo.html`

#### options.filter

- 类型：Object
- 默认值：`{}`

`filter`可以用来过滤传入url。`filter`对象，其中`key`是匹配的正则表达式，`value`是替换的字符串，支持正则表达式变量。替换的顺序与定义无关。`filter`类似于Apache中的rewrite规则

#### options.urlBasedCharset

- 类型：Object
- 默认值：`{}`

`urlBasedCharset`可针对某一个url设置响应字符集。例如：

    "charset" : "utf-8",
	"urlBasedCharset" : {"/apps/aaa.js":"gbk"}

允许在大多数情况下返回字符集为utf-8字符集的资源。但在访问/apps/aaa.js的情况下，以gbk的方式编码。 这个特性多被用来引入编码几不同的第三方脚本。

options的更多配置继承自flexcombo，请参照[flex-combo](https://npmjs.org/package/flex-combo)。

#### options.mockPath

- 类型：String
- 默认值：'mock'

本地 json(p) mock 服务脚本所在目录名（拼接到mock请求路径中），对于匹配的路径请求根据 mock 服务脚本返回值构建 json(p) 接口返回。

## 案例

#### 使用案例1

[KISSY 组件如何调试线上代码](http://blog.kissyui.com/2013/11/29/%E5%A6%82%E4%BD%95%E5%BF%AB%E9%80%9F%E5%9C%A8%E7%BA%BF%E8%B0%83%E5%BC%8F%E4%BD%A0%E7%9A%84gallery%E7%BB%84%E4%BB%B6/)

#### 使用案例2

KISSY项目构建工具和本地环境：[generator-clam](https://npmjs.org/package/generator-clam)一同使用，构建工具生成本地配置文件，形如：

	flexcombo:{
		debug:{
			options:{
				proxyHosts:'trip.taobao.com',
				target:'build/',
				proxyport:8080,
				urls:'/trip/proj/0.0.2',
				port:'80',
				servlet:'?',
				separator:',',
				charset:'utf8',
				filter:{
					'-min\\.js':'.js',
					'/go/act/\(.+\\.\)php':'/$1html',
				}
			}
		}
	}

## Thanks

- <a href="https://www.npmjs.org/package/anyproxy"><img alt="Any Proxy" width="60" src="https://i.alipayobjects.com/i/ecmng/png/201409/3NKRCRk2Uf.png_250x.png"/></a>
- <a href="http://people.apache.org/~pmuellr/weinre-docs/latest/"><img alt="weinre" width="60" src="http://gtms03.alicdn.com/tps/i3/TB1fODZGXXXXXcHXpXXTyVt9VXX-128-128.png"/></a>
- [flex-combo](https://www.npmjs.org/package/flex-combo)
- [jayli-server](https://www.npmjs.org/package/jayli-server)

history.md
### 0.1.36

- [+] `html-proxy` 引入，支持远程页面区块代理
- [+] `server` 模式下启动本地服务后自动打开本地浏览器
- [!] 清除本地 `mock` 时，`require` 的缓存，避免修改 `mock` 接口需重启服务才能刷新

### 0.1.37
- [!] bugfix for https://github.com/jayli/grunt-flexcombo/blob/master/tasks/flexcombo.js#L73，掉了个 `req` 参数传递进去
- [!] `server` 模式下启动本地服务后自动打开本地浏览器优化，带上 `-br` 启动浏览器

### 0.1.42
- [!] 启动日志显示本机 IP，方便 debug
- [!] mock 路径去除结尾的 ".do/action" 等，现在可以方便地 mock mtop 接口请求了
- [!] `demo`/`server` 模式下通过后缀 `--br` 自动打开浏览器优化

### 0.1.43
- [!] bugfix for `getLocalIp()`

### 0.2.0
- [!] 替换 reverse-proxy 为 [anyproxy](web.npm.alibaba-inc.com/package/anyproxy)
- [+] 增加 weinre 支持

### 0.2.1
- [!] 移除 `open` 支持，加入 weinre 安装检测
- [!] 移除 `--br` 打开浏览器功能

### 0.2.2
- [!] 移除遗留的 `grunt` 模块依赖（`grunt`依赖模块众多导致安装慢）

### 0.2.3
- [!] 升级 anyproxy 依赖，支持端口号可配置

### 0.2.4
- [!] 更新 `anyproxy-rule.js` 文件 `replaceRequestOption`，符合代理域名规则下的请求转发到 `flex-combo` 先做本地检查，避免本地没有导致请求 404

### 0.2.5
- [!] 更新启动提示信息，展示所有使用中的端口号

### 0.2.8
- [!] 兼容 anyproxy API 变更：`replaceServerResData` => `replaceServerResDataAsync`
- [!] 请求转发到 flexcombo 的 hostname 由 `127.0.0.1` 改为 `utils.js` 里读取本机 IP 的方法，修复移动端挂代理起 `grunt demo` 访问本机页面阻塞
- [!] Less/Sass 解析 bugfix 修复 `isUtf8` 引用

### 0.2.9
- <s>[!] 对匹配 cdn 资源判断请求 url 中是否包含 `urls` 字符串，避免代理不应该代理的线上资源</s>，有bug
- [!] 加入 anyproxy 兼容 https 请求的配置，支持 https 请求的监控

### 0.2.10
- [!] https 监控加配置项打开

### 0.2.11
- [!] bugfix for  0.2.9 prefix 判断

### 0.2.14
- [!] bugfix for mocker，修复`replaceServerResDataAsync`方法中接口 mock 回调 bug
- [!] bugfix for 空 target 时代理失效

### 0.2.15
- [!] bugfix for mocker，处理非 UTF-8 编码接口返回 bug
- [!] flex-combo 限定依赖 0.6.4

### 0.2.16
- [!] bugfix for flex-combo 转发
- [!] 加入 `?debug` 模式支持，将代理的 `-min.js` 资源请求替换为 `.js` 请求方便调试

### 0.2.17
- [!] bugfix for compatible with latest flex-combo

### 0.2.19
- [!] 加入 livereload 支持

### 0.2.20
- [!] bugfix for compatible with latest flex-combo

### 0.2.21
- [!] bugfix for proxy compatible with windows
- [!] bugfix for jsonp parse

### 0.2.22
- [!] 锁定 flex-combo 的依赖到 `~0.6.16`




test/src/base.js

/**
 * @file base.js
 * @brief Slide
 * @author jayli, bachi@taobao.com
 * @version 
 * @date 2013-01-08
 */

/*jshint smarttabs:true,browser:true,devel:true,sub:true,evil:true */

KISSY.add('a',function(S){

	"use strict";

	return {};

},{
	requires:['node','json','event']	
});



test/Gruntfile.js
module.exports = function(grunt) {
	var task = grunt.task;
    grunt.initConfig({
		// FlexCombo服务配置
		// https://npmjs.org/package/grunt-flexcombo
		flexcombo:{
			// 无线H5项目调试，可打开host配置，用法参照
			// https://speakerdeck.com/lijing00333/grunt-flexcombo
			demo:{
				options:{
					proxyport:8080,
					target:'src/',
					urls:'/test',
					proxyHosts:'demo.com',
					port:'81',
					servlet:'?',
					separator:',',
					charset:'utf-8', // 输出文件的编码
					// 默认将"-min"文件映射到源文件
					filter:{
						'-min\\.js':'.js'
					}
				}
			}
		},
		// 监听JS、CSS、LESS文件的修改
        watch: {
            'all': {
                files: ['src/**/*'],
                tasks: [ 'build' ]
            }
		},
        // 打包后压缩文件
        // 压缩文件和入口文件一一对应
        uglify: {
            options: {
                beautify: {
                    ascii_only: true
                }
            },
            base: {
                files: [{
					expand: true,
					cwd: 'src/',
					src: ['**/*.js','!**/*-min.js'],
					dest: 'src/',
					ext: '-min.js'
                }]
            }
        }
    });

    // 使用到的任务，可以增加其他任务
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-contrib-watch');
	grunt.loadNpmTasks('grunt-flexcombo');

	grunt.registerTask('build', '默认构建任务', function() {
		task.run(['uglify']);
	});

	// 启动Debug调试时的本地服务
	grunt.registerTask('demo', '开启debug模式', function() {
		task.run(['flexcombo:demo','watch:all']);
	});

    return grunt.registerTask('default', '',function(type){
		if (!type) {
			task.run(['build']);
		}
	});
};


test/src/index.js

KISSY.add('b',function(S,io){
	return {};
},{
	requires:['io']	
});


Gruntfile.js
/*
 * grunt-flexcombo
 * 
 *
 * Copyright (c) 2013 拔赤
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp'],
    },

    // Configuration to be run (and then tested).
    flexcombo: {
      default_options: {
        options: {
        },
        files: {
          'tmp/default_options': ['test/fixtures/testing', 'test/fixtures/123'],
        },
      },
	  /*
      custom_options: {
        options: {
          separator: ': ',
          punctuation: ' !!!',
        },
        files: {
          'tmp/custom_options': ['test/fixtures/testing', 'test/fixtures/123'],
        },
      },
	  */
    },


	bbb_server:{
				   options: {},
		files: {
	      'dest/default_options': ['src/testing', 'src/123'],
	  }

	}

  });




  grunt.loadNpmTasks('grunt-bbb-server');

  // Actually load this plugin's task(s).
  // grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.

  // By default, lint and run all tests.
  grunt.registerTask('default', ['bbb_server']);

};


tasks/lib/utils.js
/**
 * Created by 弘树<tiehang.lth@alibaba-inc.com> on 14-9-14.
 * util Functions
 */
var os = require('os');
var fs = require('fs');
var path = require('path');

/**
 * 控制台输出需代理的域名
 * @param proxyHosts
 */
function showProxyHosts(proxyHosts) {
    console.log('当前已代理以下域名的请求：');
    if (typeof proxyHosts != 'object') {
        return;
    }
    for (var i = 0; i < proxyHosts.length; i++) {
        console.log('  > ' + green('http://' + proxyHosts[i] + '/'));
    }
}

function consoleColor(str, num) {
    num = num || '32';
    return "\033[" + num + "m" + str + "\033[0m";
}

function green(str) {
    return consoleColor(str, 32);
}

function yellow(str) {
    return consoleColor(str, 33);
}

function red(str) {
    return consoleColor(str, 31);
}

function blue(str) {
    return consoleColor(str, 34);
}

function log(statCode, url, err) {
    var logStr = blue(statCode) + ' - ' + url;
    if (err)
        logStr += ' - ' + red(err);
    console.log(logStr);
}

/**
 * 读取给定路径下的文件信息，生成 html 展示页面
 * @param dir {String}
 * @returns {String}
 */
function getDirFiles(dir) {
    var files = fs.readdirSync(dir);
    var res_f = [];
    var res_d = [];
    var r = '<!DOCTYPE html><html><head><meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1"></head><body>';
    files.forEach(function (file) {
        var stat = fs.lstatSync(path.resolve(dir, file));

        if (!stat.isDirectory()) {
            res_f.push(file);
        } else {
            res_d.push(file);
        }
    });

    r += '<p><img src="http://img02.taobaocdn.com/tps/i2/T1WNlnFadjXXaSQP_X-16-16.png" /> <a href="../">parent dir</a></p><hr size=1 />';

    res_d.forEach(function (file) {
        r += '<p><img src="http://img03.taobaocdn.com/tps/i3/T1nHRTFmNXXXaSQP_X-16-16.png" /> <a href="' + file + '/">' + file + '</a></p>';
    });

    res_f.forEach(function (file) {
        r += '<p><img src="http://img02.taobaocdn.com/tps/i2/T1Y7tPFg8eXXaSQP_X-16-16.png" /> <a href="' + file + '">' + file + '</a></p>';
    });

    r += '</body></html>';

    return r;
}

/**
 * 给定路径是否为目录
 * @param dir {String}
 * @returns {Boleaan}
 */
function isDir(dir) {
    if (fs.existsSync(dir)) {
        var stat = fs.lstatSync(dir);
        return stat.isDirectory();
    } else {
        return false;
    }
}

/**
 * 给定路径是否为文件
 * @param dir {String}
 * @returns {Boolean}
 */
function isFile(dir) {
    if (fs.existsSync(dir)) {
        var stat = fs.lstatSync(dir);
        return stat.isFile();
    } else {
        return false;
    }
}

/**
 * 读取本机 IP 地址（IPv4）
 * @returns {String}
 */
function getLocalIp() {
    var ifaces = os.networkInterfaces();
    var lookupIpAddress = null;
    for (var dev in ifaces) {
        if (dev != "en1" && dev != "en0") {
            continue;
        }
        ifaces[dev].forEach(function (details) {
            if (details.family == 'IPv4') {
                lookupIpAddress = details.address;
                return;
            }
        });
    }
    return lookupIpAddress || '127.0.0.1';
}

/**
 * 从 jsonp 格式数据解析出 json
 * @param jsonpRet
 * @returns {Object}
 */
function parseJsonp(jsonpRet) {

    if (jsonpRet) {

        var ret = jsonpRet.replace(/^\w*\((.*)\)$/, '$1');
        return JSON.parse(ret);
    }

    return {};

}


exports.showProxyHosts = showProxyHosts;
exports.isDir = isDir;
exports.isFile = isFile;

exports.green = green;
exports.yellow = yellow;
exports.red = red;
exports.blue = blue;

exports.log = log;

exports.getDirFiles = getDirFiles;
exports.getLocalIp = getLocalIp;
exports.parseJsonp = parseJsonp;

tasks/flexcombo.js
/**
 * grunt-flexcombo
 *
 * Copyright (c) 2013 拔赤
 * Licensed under the MIT license.
 */

var path = require('path');
var http = require('http');
var url = require('url');

var flexCombo = require('flex-combo');
var jayli = require('jayli-server');
var iconv = require('iconv-lite');
var proxy = require("anyproxy");
require('shelljs/global');

var ProxyRule = require('./lib/anyproxy-rule');
var utils = require('./lib/utils');

// 默认的 liveReload 端口号
var DEFAULT_LIVERELOADPORT = 35729;

module.exports = function (grunt) {

    // Please see the Grunt documentation for more information regarding task
    // creation: http://gruntjs.com/creating-tasks

    grunt.registerMultiTask('flexcombo', 'Start FlexCombo.', function () {
        // Merge task-specific and/or target-specific options with these defaults.
        var options = this.options();

        var that = this;
        var pwd = process.cwd();
        var port = options.port || '80';
        var proxyport = options.proxyport || 8080;

        // anyproxy 端口配置
        var webConfigPort = options.webConfigPort || 8082;
        var webPort = options.webPort || 8002;
        var socketPort = options.socketPort || 8003;

        // 是否启动 weinre
        var startWeinre = options.startWeinre;
        var weinrePort = options.weinrePort || 9090;
        var prefix = options.urls;
        var target = options.target;
        var localPath = options.target;
        var filter = options.filter || {};
        var livereload = options.livereload;	// 是否需要自动刷新，配置为自动刷新端口号 or true
        var livereloadPort = ((livereload === true) ? DEFAULT_LIVERELOADPORT : livereload); //livereload工作端口号
        var needHttps = options.needHttps;      // 是否需要启用 https 监控
        var localIpAddr = utils.getLocalIp();

        // 代理接口和页面的默认配置
        var proxyIfPageConfig = options.proxy || {
                interface: {
                    hosts: [],
                    script: 'proxy/interface.js'
                },
                webpage: {
                    urls: [],
                    script: 'proxy/webpage.js'
                }
            };

        // 代理主机配置项
        var proxyHosts = [];
        if (typeof options.proxyHosts === 'object') {
            proxyHosts = options.proxyHosts;
        } else if (typeof options.proxyHosts === 'string') {
            proxyHosts = [options.proxyHosts];
        }

        // 初始化 flex-combo
        var obj = {};
        obj[options.urls] = options.target;
        for (var i in obj) {
            obj[i] = path.resolve(pwd, obj[i]);
        }
        var comboInst = flexCombo(process.cwd(), obj, options);

        if (needHttps) {
            //create cert when you want to use https features
            //please manually trust this rootCA when it is the first time you run it
            !proxy.isRootCAFileExists() && proxy.generateRootCA();
        }

        // 初始化 anyproxy
        new proxy.proxyServer({
            type: "http",
            port: proxyport,
            // port for web interface
            webPort: webPort,
            // internal port for web socket, replace this when it is conflict with your own service
            socketPort: socketPort,
            // internal port for web config(beta), replace this when it is conflict with your own service
            webConfigPort: webConfigPort,
            hostname: "localhost",
            rule: ProxyRule({
                prefix: prefix,
                filter: filter,
                port: port,
                proxyHosts: proxyHosts,
                target: target,
                proxy: proxyIfPageConfig,
                needHttps: needHttps,
                pwd: pwd
            })
        });

        // 启动本地服务
        http.createServer(function (req, res) {
            comboInst(req, res, function () {

                /**
                 * flex-combo 请求未匹配上时的额外处理
                 */
                // get true path
                var parsedReqUrl = url.parse(req.url, true);
                var truePath = path.resolve(pwd, localPath, req.url.replace('http://' + req.headers.host, '').replace(prefix, './'));
                truePath = truePath.replace(/\?.*$/, '');
                truePath = truePath.replace(/#.*$/, '').replace('http:/' + req.headers.host, '');
                var removedPrefixPath = parsedReqUrl.pathname.replace(prefix, '');
                if (utils.isDir(truePath)) {

                    /**
                     * 如果请求的是目录路径，展示该目录下的文件和文件夹列表
                     */
                    if (!/\/$/.test(req.url)) {
                        res.writeHead(302, {'Content-Type': 'text/html'});
                        res.end('<script>window.location.href += "/";</script>');
                        return;
                    }
                    res.writeHead(200, {'Content-Type': 'text/html'});
                    var r = utils.getDirFiles(truePath);
                    r += '<style>' +
                    'p {' +
                    'font-family:Tahoma;' +
                    'font-size:1em;' +
                    'line-height:1em;' +
                    'margin:9px 10px' +
                    '}' +
                    '</style>';
                    res.end(r);
                } else if (utils.isFile(truePath)) {

                    /**
                     * 如果请求的是其他文件，转入 jayli-server 静态文件服务
                     * 对于页面（html）提供 ssi、juicer mock、php 语法兼容等服务
                     */
                    jayli
                        .deliver('.', req, res, truePath)
                        .before(function () {
                        })
                        .after(function (statCode, filePath, fileBuffer, encoding) {
                            utils.log(statCode, req.url);

                            var decodedFileContent = iconv.decode(fileBuffer, encoding);
                            // inject weinre
                            if (startWeinre) {

                                // inject weinre script
                                var weinreScriptPath = "http://" + localIpAddr + ":" + weinrePort
                                    + "/target/target-script-min.js#anonymous";
                                decodedFileContent = decodedFileContent.replace('</body>', '<script src="' + weinreScriptPath + '"></script></body>');
                            }

                            // inject live-reload
                            if (livereload) {

                                // inject live-reload script
                                var liveReloadScriptPath = "http://" + localIpAddr + ":"
                                    + ((livereload === true) ? DEFAULT_LIVERELOADPORT : livereload)
                                    + "/livereload.js?snipver=1";
                                decodedFileContent = decodedFileContent.replace('</body>', '<script src="' + liveReloadScriptPath + '"></script></body>');
                            }

                            fileBuffer = iconv.encode(decodedFileContent, encoding);
                            return fileBuffer;
                        });
                } else {
                    utils.log(404, req.url, 'Not found');
                    res.writeHead(404, {'Content-Type': 'text/plain'});
                    res.end("Error 404: " + req.url);
                }
            });
        }).listen(port, function () {
            // 从sudo降权，避免watch build后的文件为root权限
            if (process.setgid && process.setuid) {

                var env = process.env,
                    uid = parseInt(env['SUDO_UID'] || process.getuid(), 10),
                    gid = parseInt(env['SUDO_GID'] || process.getgid(), 10);

                process.setgid(gid);
                process.setuid(uid);

            }
        });

        console.log('\nAnyproxy web config service running at ' + utils.blue('http://127.0.0.1:' + webConfigPort));
        console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('webConfigPort') + ' 更改端口号');
        console.log('\nAnyproxy web monitor service running at ' + utils.blue('http://127.0.0.1:' + webPort));
        console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('webPort') + ' 更改端口号');
        console.log('\nAnyproxy web socket service running at ' + utils.blue('http://127.0.0.1:' + socketPort));
        console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('socketPort') + ' 更改端口号');
        console.log('\nFlex Combo Server running at ' + utils.blue('http://127.0.0.1:' + port));
        console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('port') + ' 更改端口号');

        if(livereload && livereloadPort) {
            console.log('\nLiveReload running at ' + utils.blue('http://127.0.0.1:' + livereloadPort));
            console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('livereload') + ' 更改端口号');
        }

        console.log('\n 请为你的浏览器或移动设备配置 HTTP 代理: '
        + utils.green('127.0.0.1(浏览器) / ' + localIpAddr + '(移动设备)' + ' : ' + proxyport));
        console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('proxyport') + ' 更改端口号');

        console.log('\nHelp: ' + utils.blue('http://groups.alidemo.cn/trip-tools/clam-doc/intro/dev-debug.html'));

        utils.showProxyHosts(proxyHosts);

        console.log('请访问以下链接查看 anyproxy 网络监控：');
        console.log('    ' + utils.green('http://localhost:' + webPort + ''));

        // 打开 weiner
        if (startWeinre) {

            var checkWeinre = exec('weinre -h', {silent: true});
            if (checkWeinre.code === 127) {
                // weinre 未安装
                console.log(utils.yellow('您尚未安装 weinre，可执行 tnpm install -g weinre 安装 weinre'));
            } else {
                console.log('请访问以下链接查看 weinre 控制面板：');
                console.log('    ' + utils.green('http://' + localIpAddr + ':' + weinrePort + '/client/#anonymous'));
                console.log('\n Weinre service running at ' + utils.blue('http://127.0.0.1:' + weinrePort));
                console.log('  > 可通过 flexocombo 配置项 ' + utils.yellow('weinrePort') + ' 更改端口号');

                exec('weinre --boundHost ' + localIpAddr + ' --httpPort ' + weinrePort, function (code, output) {

                });
            }

        }
        console.log(utils.yellow('如遇到 Error: listen EADDRINUSE 报错提示，请检查以上端口号是否被占用。'));

    });

};
